### PR TITLE
WIP: Fix error in data create

### PR DESCRIPTION
**Pull request by mstajdohar**
To be merged with @dblenkus's branch.

### CODE REVIEW

File: resolwe/flow/models.py [hunk 1]
```diff
@@ -420,19 +420,17 @@ def save(self, *args, **kwargs):
...
+                        unique_slug = "{} {}".format(self.name, i)
                         i += 1
 
                     collection = Collection.objects.create(
                         contributor=self.contributor,
                         descriptor_schema=des_schema,
-                        name=unique_col_name,
-                        slug=slugify(unique_col_name),
+                        name=self.name,
+                        slug=slugify(unique_slug),
```
Comment (mstajdohar) [inline resolwe/flow/models.py]: @dblenkus maybe use [django_unique_slugify](https://github.com/defcube/django-unique-slugify) or similar.

Comment (mstajdohar) [inline resolwe/flow/models.py]: If name not given, the template should be defined in the process object. We should add a field `name_data_default` to processes. Does anyone (@jkokosar, @kostko, @dblenkus) have a better suggestion for the field name?

File: resolwe/flow/views.py [hunk 2]
```diff
@@ -135,23 +135,35 @@ class ResolweCreateDataModelMixin(ResolweCreateModelMixin):
...
-            return Response({'collections': ['Invalid pk "{}" - object does not exist.'.format(collections[0])]},
-                            status=status.HTTP_400_BAD_REQUEST)
-
-        if not request.user.has_perm('add_collection', obj=collection):
-            if request.user.is_authenticated():
-                raise exceptions.PermissionDenied
-            else:
-                raise exceptions.NotFound
+        
+        for collection_id in collections:
```
Comment (mstajdohar) [inline resolwe/flow/views.py]: Collections are allowed in REST API, but the data object is not added to them. @dblenkus?

Comment (mstajdohar) [inline resolwe/flow/views.py]: I have implemented the REST API so that users do not have to list collections, but they can if they want to.

Comment (mstajdohar) [inline resolwe/flow/views.py]: Probably not the most efficient implementation.

Comment (dblenkus) [inline resolwe/flow/views.py]: I've already fixed this on my branch. The problem here is how to construct response if user has permissions on all collections except one (or maybe one does not exist).
Quick solution is to disable setting any collections on create and then make separate requests.

Comment (dblenkus) [inline resolwe/flow/views.py]: One more question is how to make requests for adding data objects to collections. Do you have to post list of all collections? In this case you can (accidentally override updated list on server). The more secure way would be to create special endpoints, ie. `collection/<id>/add_data` and `collection/<id>/remove_data`.
What do you think @mstajdohar?

Comment (mstajdohar) [inline resolwe/flow/views.py]: I vote to fail the whole transaction.

File: resolwe/flow/models.py [hunk 3]
```diff
@@ -420,19 +420,17 @@ def save(self, *args, **kwargs):
...
                 else:
                     des_schema = DescriptorSchema.objects.get(slug=self.process.flow_collection)
-
-                    col_name = "{}_{}".format(self.descriptor['name'], self.flow_collection)
-                    unique_col_name = col_name
+                    unique_slug = self.name
                     i = 1
-                    while Collection.objects.filter(name=unique_col_name).exists():
-                        unique_col_name = "{}_{}".format(col_name, i)
+                    while Collection.objects.filter(slug=slugify(unique_slug)).exists():
```
Comment (dblenkus) [inline resolwe/flow/models.py]: +1

File: resolwe/flow/models.py (hunk 1 above)
Comment (dblenkus) [inline resolwe/flow/models.py]: Thanks, I will check it.

File: resolwe/flow/views.py (hunk 2 above)
Comment (mstajdohar) [inline resolwe/flow/views.py]: I do not expect this feature (creating new data on multiple collections simultaneously) will be used much.

Comment (dblenkus) [inline resolwe/flow/views.py]: I agree that failing whole transaction is easiest solution. But do we want to return just first error or all of them?
We have to thing what to support and what not. If something won't be used much and is to complicated to think through all possibilities then I don't see a reason not to disable it.

**Comment by mstajdohar** [comment]
@dblenkus, maybe you should merge this with your branch and do a unified PR.

File: resolwe/flow/views.py (hunk 2 above)
Comment (mstajdohar) [inline resolwe/flow/views.py]: The web devs (@kostko, @hadalin) should tell how they would like to append data objects to collections at create and what they would want in return. Adding a data object to a one collection is definitely used in GenBoard. The zero collections alternative is used by @JureZmrzlikar in resolwe-bio-py. I do not expect a case with multiple collections in the near future. But I do not like exceptions and it is quite ugly to me to support just one element in an array.

File: resolwe/flow/models.py (hunk 1 above)
Comment (kostko) [inline resolwe/flow/models.py]: How was the name generated on the old platform?

Comment (dblenkus) [inline resolwe/flow/models.py]: Template was defined in static schema (now descriptor schema) and generated in `render_static` function. We also have `render_descriptor` function in resolwe, so we probably just have to copy name from `descriptor.name` to `name`.

Comment (dblenkus) [inline resolwe/flow/models.py]: This is for the name of `Data` object. But here we generate name of `sample`, which is a concept that wasn't supported on old platform.

**Comment by tjanez** [comment]
@mstajdohar, @dblenkus: Please, rebase this or the other branch on top of `master` branch and add a CHANGELOG entry.

**Comment by mstajdohar** [comment]
I think @dblenkus will change this completely with his updated code.

**Pull request closed without merging**